runtime: Make csi-directvolume-kata works well with runtime

Apokleos opened the pull request:
As csi-directvolume-kata will use raw file on the host as the backend of a virtio-blk block device, and it is treated as just a regular file, which cause runtime label it unsupported device type. To correct such case, this patch supports it.
Another important factor related to the issue #11296, as it said that, the Options with bind does cause it fail. But in my test with runtime-rs, it works well. So in this patch, I just remove the related Option assignment from the m.Option

Fixes #11296

sprt commented:
> As csi-directvolume-kata will use raw file on the host as the backend of a virtio-blk block device, and it is treated as just a regular file

Doesn't this require VMM support? i.e. this change may work with QEMU but not CH?

Apokleos commented:
> > As csi-directvolume-kata will use raw file on the host as the backend of a virtio-blk block device, and it is treated as just a regular file
> 
> Doesn't this require VMM support? i.e. this change may work with QEMU but not CH?
>
Hey @sprt Thx.
Its goal is to support all vmms which support rawfiles, based on this, I think it should work on the CLH, if not, some more work are needed. Have you tested it in clh?

sprt commented:
@Apokleos I vaguely remember raw files not working with CH but I hadn't tried this change, instead I had modified the driver to support loop devices: https://github.com/kata-containers/kata-containers/pull/10559/commits/29fc95b169cf235064e6d4627c88db1fbed744a5

@fidencio Have you tested this change with CH?

Xynnn007 commented:
Thanks @Apokleos . I test this locally with kata 3.13 + qemu-tdx and works well
```yaml
kind: Pod
apiVersion: v1
metadata:
  name: kata-driectvol-01
spec:
  runtimeClassName: kata-qemu-tdx
  containers:
    - name: first-container
      image: xxxx*:ubuntu-22.04
      volumeMounts:
      - mountPath: "/data"
        name: kata-driectvol0-volume
      command: [ "sleep", "1000000" ]
  volumes:
    - name: kata-driectvol0-volume
      persistentVolumeClaim:
        claimName: csi-directvolume-pvc # defined in csi-pvc.yaml

```

logs

```
Events:
  Type    Reason     Age   From               Message
  ----    ------     ----  ----               -------
  Normal  Scheduled  16s   default-scheduler  Successfully assigned default/kata-driectvol-01 to intel-g220-ba
  Normal  Pulled     11s   kubelet            Container image "xxxx*:ubuntu-22.04" already present on machine
  Normal  Created    11s   kubelet            Created container: first-container
  Normal  Started    7s    kubelet            Started container first-container
root@intel-G220-BA:~/tdx/development/kata-containers/src/tools/csi-kata-directvolume# kubectl get pods
NAME                READY   STATUS    RESTARTS   AGE
kata-driectvol-01   1/1     Running   0          21s
root@intel-G220-BA:~/tdx/development/kata-containers/src/tools/csi-kata-directvolume# kubectl exec -it kata-driectvol-01 -- bash
root@kata-driectvol-01:/# ls /data
lost+found
```

But with `runtimeClassName` as `kata`, it fails

```yaml
kind: Pod
apiVersion: v1
metadata:
  name: kata-driectvol-01
spec:
  runtimeClassName: kata
  containers:
    - name: first-container
      image: xxxx*:ubuntu-22.04
      volumeMounts:
      - mountPath: "/data"
        name: kata-driectvol0-volume
      command: [ "sleep", "1000000" ]
  volumes:
    - name: kata-driectvol0-volume
      persistentVolumeClaim:
        claimName: csi-directvolume-pvc # defined in csi-pvc.yaml
```

logs
```
  Normal   Scheduled  14s                default-scheduler  Successfully assigned default/kata-driectvol-01 to intel-g220-ba
  Normal   Pulled     13s (x2 over 14s)  kubelet            Container image "xxxx*:ubuntu-22.04" already present on machine
  Normal   Created    13s (x2 over 14s)  kubelet            Created container: first-container
  Warning  Failed     12s (x2 over 13s)  kubelet            Error: failed to create containerd task: failed to create shim task: No such file or directory (os error 2)

Stack backtrace:
   0: <unknown>
   1: <unknown>
   2: <unknown>
   3: <unknown>
   4: <unknown>
   5: <unknown>
   6: <unknown>
   7: <unknown>

Stack backtrace:
   0: <unknown>
   1: <unknown>
   2: <unknown>
   3: <unknown>
   4: <unknown>
   5: <unknown>
   6: <unknown>
   7: <unknown>
   8: <unknown>
   9: <unknown>
  10: <unknown>
  11: <unknown>
  12: <unknown>
  13: <unknown>: unknown
  Warning  BackOff  11s (x2 over 12s)  kubelet  Back-off restarting failed container first-container in pod kata-driectvol-01_default(bdb84870-868b-4673-8296-d2df57b7c309)
```

Apokleos commented:
> @Apokleos I vaguely remember raw files not working with CH but I hadn't tried this change, instead I had modified the driver to support loop devices: [29fc95b](https://github.com/kata-containers/kata-containers/commit/29fc95b169cf235064e6d4627c88db1fbed744a5)
> 
> @fidencio Have you tested this change with CH?
>
With this patch, CH with csi-direct-volume works well as below:
```
root@pek001:/home/pek001/katadev/NYDUS# kubectl exec -ti kata-directtvol-clh-test -- lsblk
NAME      MAJ:MIN RM  SIZE RO TYPE MOUNTPOINTS
vda       254:0    0    1G  0 disk /data
pmem0     259:0    0  254M  0 disk 
`-pmem0p1 259:1    0  253M  0 part 
root@pek001:/home/pek001/katadev/NYDUS# crictl pods |grep kata
776274c52efe0       58 seconds ago      Ready               kata-directtvol-clh-test          default             0                   kata
d572171189563       4 days ago          Ready               kata-driectvol-01                 default             0                   kata
f3e14a109a753       10 days ago         Ready               csi-kata-directvol-plugin-g8tdv   kata-directvolume   41                  (default)
root@pek001:/home/pek001/katadev/NYDUS# ps -ef |grep 776274c52efe0
root     3851507       1  0 09:56 ?        00:00:00 /opt/kata/bin/containerd-shim-kata-v2 -namespace k8s.io -address /run/containerd/containerd.sock -publish-binary  -id 776274c52efe066d652c458190736fe8b94133288305a9c72883462afaf94cff -debug
root     3851515 3851507  0 09:56 ?        00:00:00 /opt/kata/libexec/virtiofsd --syslog --cache=auto --shared-dir=/run/kata-containers/shared/sandboxes/776274c52efe066d652c458190736fe8b94133288305a9c72883462afaf94cff/shared --fd=3 --thread-pool-size=1 --announce-submounts
root     3851516 3851507  1 09:56 ?        00:00:00 /opt/kata/bin/cloud-hypervisor --api-socket /run/vc/vm/776274c52efe066d652c458190736fe8b94133288305a9c72883462afaf94cff/clh-api.sock
root     3851520 3851515  0 09:56 ?        00:00:00 /opt/kata/libexec/virtiofsd --syslog --cache=auto --shared-dir=/run/kata-containers/shared/sandboxes/776274c52efe066d652c458190736fe8b94133288305a9c72883462afaf94cff/shared --fd=3 --thread-pool-size=1 --announce-submounts
root     3851889 3845186  0 09:57 pts/2    00:00:00 grep --color=auto 776274c52efe0
```

sprt commented on src/runtime/pkg/resourcecontrol/utils.go:
```diff
@@ -36,9 +36,10 @@ func DeviceToCgroupDeviceRule(device string) (*devices.Rule, error) {
 	case unix.S_IFCHR:
 		deviceRule.Type = 'c'
 	case unix.S_IFBLK:
+	case unix.S_IFREG:
 		deviceRule.Type = 'b'
```
In Go, case statements do NOT fall through, meaning this code is now equivalent to below - is this intended?
```golang
	case unix.S_IFBLK:
        // Do nothing, deviceRule.type is NOT set
	case unix.S_IFREG:
		deviceRule.Type = 'b'
```

sprt reviewed (commented):
Thank you @Apokleos. I have one comment you probably want to look at.

Other than that, it's unfortunate we don't have a new test for this. How about I cherry-pick this commit into #10559 and we leverage the test there instead?

sprt commented on src/runtime/virtcontainers/container.go:
```diff
@@ -639,7 +646,9 @@ func (c *Container) createBlockDevices(ctx context.Context) error {
-			c.mounts[i].Options = mntInfo.Options
+			c.mounts[i].Options = slices.DeleteFunc(mntInfo.Options, func(opt string) bool {
+				return opt == "bind"
+			})
```
@Apokleos Per https://github.com/kata-containers/kata-containers/issues/11296#issuecomment-2898087290 we should probably fix the driver instead?

Otherwise I've successfully tested this PR manually. 🙂

Apokleos committed:
csi-kata-directvolume: Remove 'bind' option from CSI direct volume mounts

The `csi-kata-directvolume` driver was incorrectly including "bind" in
direct volume mount options.

This caused agent mount failures as `MS_BIND` is unsuitable for block
devices.

This commit removes the "bind" option to ensure correct direct volume
mounting.

Fixes #11296

Signed-off-by: Alex Lyn <alex.lyn@antgroup.com>

Apokleos committed:
runtime: Make csi-directvolume-kata works well within runtime

As csi-directvolume-kata will use raw file on the host as the
backend of a virtio-blk block device, and it is treated as just a
regular file, which cause runtime label it unsupported device type.
To correct such case, this patch supports it.

Signed-off-by: Alex Lyn <alex.lyn@antgroup.com>

Apokleos committed:
runtime: fix unit test for regular file handling

Previously, the test expected an error and a nil device for regular
files, causing test failures.

This change updates the test case to aligns the unit test with actual
function behavior and prevents false failures.

Signed-off-by: Alex Lyn <alex.lyn@antgroup.com>

sprt commented:
Superseded by #10559.
